src/amqp/connection.js

```diff
@@ -48,9 +48,7 @@ function getOption (opts, key, alt) {
 }
 
 function getUri (protocol, user, pass, server, port, vhost, heartbeat) {
-  return protocol + user + ':' + pass +
-    '@' + server + ':' + port + '/' + vhost +
-    '?heartbeat=' + heartbeat;
+  return `${protocol}://${user}:${pass}@${server}:${port}/${vhost}?heartbeat=${heartbeat}`;
 }
```

Comment:
I think this is the right direction.  At this point we should just use URL, and not encode things ourselves.

something like 

```suggestion
function getUri(protocol, user, pass, server, port, vhost, heartbeat) {
  const url = new URL(`${protocol}://${server}`);
  url.username = user;
  url.password = pass;
  url.port = port;
  url.pathname = vhost;
  url.searchParams.append("heartbeat", heartbeat);
  return url.toString();
}
```

or even one step further and just inline this in `getNextUri` given `getUri` is not used anywhere else.

Reply:
I think string concatenation is the easiest way to do it. `amqplib` does it natively and it avoids having the previous issue of double escaping. Edit: Please see my justification in my [other response](https://github.com/Foo-Foo-MQ/foo-foo-mq/pull/56#discussion_r1555924430)